[Data][Flaky] `test_limit_pushdown_conservative` fails due to non-deterministic task ordering

## Test Name
`test_limit_pushdown_conservative`

## Test Location
`python/ray/data/tests/test_execution_optimizer_limit_pushdown.py:82`

## Issue Description
The test fails intermittently when checking limit pushdown behavior because it assumes a specific ordering of results that is not guaranteed by the Ray Data interface.

## Root Cause
The test creates a dataset with `override_num_blocks=100`, applies operations, limits to 5 rows, and expects the result to be `[{"id": 0}, {"id": 1}, {"id": 2}, {"id": 3}, {"id": 4}]`. However, when tasks finish out of order, the actual result can be `[{"id": 0}, {"id": 2}, {"id": 3}, {"id": 4}, {"id": 1}]`, causing the assertion to fail.

The issue is that `take_all()` returns rows in the order they are produced by the distributed execution, which depends on task scheduling and completion timing.

## Example Failure
```
AssertionError: assert [{'id': 0}, {'id': 2}, {'id': 3}, {'id': 4}, {'id': 1}] == [{'id': 0}, {'id': 1}, {'id': 2}, {'id': 3}, {'id': 4}]
  At index 1 diff: {'id': 2} != {'id': 1}
```

## Proposed Fix
Rewrite the assertion in a way that avoids the assumption that results are returned in a specific order. This affects multiple test cases in the function that use `_check_valid_plan_and_result()`.

## Additional Context
This issue occurs because the test assumes deterministic ordering from distributed operations, which can vary based on task scheduling and completion timing.

--- a/python/ray/data/tests/test_execution_optimizer_limit_pushdown.py
+++ b/python/ray/data/tests/test_execution_optimizer_limit_pushdown.py
@@ -19,8 +19,21 @@ def _check_valid_plan_and_result(
     expected_plan: Plan,
     expected_result: List[Dict[str, Any]],
     expected_physical_plan_ops=None,
+    check_ordering=True,
 ):
-    assert ds.take_all() == expected_result
+    actual_result = ds.take_all()
+    if check_ordering:
+        assert actual_result == expected_result
+    else:
+        # When ordering doesn't matter, compare as multisets.
+        from collections import Counter
+
+        def to_hashable(d):
+            return tuple(sorted(d.items()))
+
+        assert Counter(map(to_hashable, actual_result)) == Counter(
+            map(to_hashable, expected_result)
+        )
     assert ds._plan._logical_plan.dag.dag_str == expected_plan
 
     expected_physical_plan_ops = expected_physical_plan_ops or []
@@ -83,6 +96,7 @@ def test_limit_pushdown_conservative(ray_start_regular_shared_2_cpus):
         ds,
         "Read[ReadRange] -> Limit[limit=5] -> Project[Project]",
         [{"id": i} for i in range(5)],
+        check_ordering=False,
     )
 
     # Test 6: Limit should stop at Sort operations (AllToAll)

--- a/python/ray/data/tests/test_execution_optimizer_limit_pushdown.py
+++ b/python/ray/data/tests/test_execution_optimizer_limit_pushdown.py
@@ -7,6 +7,7 @@ import pytest
 import ray
 from ray.data import Dataset
 from ray.data._internal.logical.interfaces import Plan
+from ray.data._internal.util import rows_same
 from ray.data.block import BlockMetadata
 from ray.data.datasource import Datasource
 from ray.data.datasource.datasource import ReadTask

--- a/python/ray/data/tests/test_execution_optimizer_limit_pushdown.py
+++ b/python/ray/data/tests/test_execution_optimizer_limit_pushdown.py
@@ -25,7 +25,6 @@ def _check_valid_plan_and_result(
     if check_ordering:
         assert actual_result == expected_result
     else:
-        # When ordering doesn't matter, compare as multisets.
         assert rows_same(pd.DataFrame(actual_result), pd.DataFrame(expected_result))
     assert ds._plan._logical_plan.dag.dag_str == expected_plan
 

--- a/python/ray/data/tests/test_execution_optimizer_limit_pushdown.py
+++ b/python/ray/data/tests/test_execution_optimizer_limit_pushdown.py
@@ -26,14 +26,7 @@ def _check_valid_plan_and_result(
         assert actual_result == expected_result
     else:
         # When ordering doesn't matter, compare as multisets.
-        from collections import Counter
-
-        def to_hashable(d):
-            return tuple(sorted(d.items()))
-
-        assert Counter(map(to_hashable, actual_result)) == Counter(
-            map(to_hashable, expected_result)
-        )
+        assert rows_same(pd.DataFrame(actual_result), pd.DataFrame(expected_result))
     assert ds._plan._logical_plan.dag.dag_str == expected_plan
 
     expected_physical_plan_ops = expected_physical_plan_ops or []

